### PR TITLE
nixos/acme: add s3Bucket option

### PR DESCRIPTION
**Pull request by yu-re-ka**
## Description of changes

Use case: serving on an anycasted IP, with the well-known/acme-challenge being backed by an S3 bucket.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/security/acme/default.nix [hunk 1]
```diff
@@ -929,32 +940,23 @@ in {
...
+          assertion = let
+            boolToInt = x: if false then 0 else 1;
+            sum = l: lib.fold (x: y: x + y) 0 l;
+          in
+            sum (map (x: boolToInt (x != null)) [
+              data.dnsProvider
+              data.webroot
+              data.listenHTTP
+              data.s3Bucket
+            ]) != 1;
```
Comment (emilylange) [inline nixos/modules/security/acme/default.nix:952]: I find the following much easier to read:

```suggestion
          assertion = lib.length (lib.filter (x: x != null) [
            data.dnsProvider
            data.webroot
            data.listenHTTP
            data.s3Bucket
          ]) != 1;
```

File: nixos/modules/security/acme/default.nix [hunk 2]
```diff
@@ -363,12 +365,12 @@ let
         ];
 
         # Only try loading the environmentFile if the dns challenge is enabled
```
Comment (emilylange) [inline nixos/modules/security/acme/default.nix:367]: Please update the comment or remove it :)

**Review by emilylange — CHANGES_REQUESTED** [review]
This is AWS s3 exclusive, as `lego` does not expose a way to specify a custom s3 endpoint, right?
https://github.com/go-acme/lego/pull/1970/files

So we don't have a way to test this (which is fine).

**Comment by yu-re-ka** [comment]
> This is AWS s3 exclusive, as lego does not expose a way to specify a custom s3 endpoint, right?

Yes, I have a patch locally:

```diff
File: ../nixfiles/pkgs/lego.diff
diff --git a/providers/http/s3/s3.go b/providers/http/s3/s3.go
index d8fc4246..84bf3c46 100644
--- a/providers/http/s3/s3.go
+++ b/providers/http/s3/s3.go
@@ -5,6 +5,7 @@ import (
    "bytes"
    "context"
    "fmt"
+   "os"
    "strings"
 
    "github.com/aws/aws-sdk-go-v2/aws"
@@ -33,7 +34,11 @@ func NewHTTPProvider(bucket string) (*HTTPProvider, error) {
        return nil, fmt.Errorf("s3: unable to create AWS config: %w", err)
    }
 
-   client := s3.NewFromConfig(cfg)
+   client := s3.NewFromConfig(cfg, func(o *s3.Options) {
+       o.EndpointResolver = s3.EndpointResolverFromURL(os.Getenv("AWS_ENDPOINT_URL"), func (e *aws.Endpoint) {
+           e.HostnameImmutable = true
+       })
+   })
 
    return &HTTPProvider{
        bucket: bucket,

**Review by emilylange — APPROVED** [review]

**Comment by vcunat** [comment]
For reference also on PR, this blocks evaluation and thus channels (for `unstable-small` already).
```
nixos.tests.acme.aarch64-linux: error:
       … while calling the 'head' builtin

         at /nix/store/8vm2qc8gqki6i8avyrwy3sjhvc100nib-source/lib/attrsets.nix:850:11:

          849|         || pred here (elemAt values 1) (head values) then
          850|           head values
             |           ^
          851|         else

       … while calling the 'deepSeq' builtin

         at /nix/store/8vm2qc8gqki6i8avyrwy3sjhvc100nib-source/lib/customisation.nix:286:7:

          285|     in if drv == null then null else
          286|       lib.deepSeq drv' drv';
             |       ^
          287|

       (stack trace truncated; use '--show-trace' to show the full trace)

       error:
       Failed assertions:
       - Exactly one of the options
       `security.acme.certs.http.example.test.dnsProvider`,
       `security.acme.certs.http.example.test.webroot`,
       `security.acme.certs.http.example.test.listenHTTP` and
       `security.acme.certs.http.example.test.s3Bucket`
       is required.
```

**Pull request merged**